Mathlib/Data/Matrix/ColumnRowPartitioned.lean

```diff
@@ -0,0 +1,332 @@
+  funext i j
+  cases' i with i i;
+  all_goals (cases' j with j j)
+  all_goals ( simp only [mul_fromColumns, fromRows_mul, fromColumns_apply_inl,
+    fromColumns_apply_inr, fromRows_apply_inl, fromRows_apply_inr, fromBlocks_apply₁₁,
+    fromBlocks_apply₁₂, fromBlocks_apply₂₁, fromBlocks_apply₂₂] )
+
+/-- A column partitioned matrix mulitplied by a row partitioned matrix gives the sum of the "outer"
+products of the block matrices -/
+lemma fromColumns_mul_fromRows (A₁ : Matrix m n₁ R) (A₂ : Matrix m n₂ R)
+    (B₁ : Matrix n₁ n R) (B₂ : Matrix n₂ n R) :
+    fromColumns A₁ A₂ ⬝ fromRows B₁ B₂ = (A₁ ⬝ B₁ + A₂ ⬝ B₂) := by
+  funext i j
+  rw [fromRows, fromColumns]
+  simp only [add_apply, mul_apply, of_apply, Fintype.sum_sum_type, Sum.elim_inl, Sum.elim_inr]
+
+/-- A column partitioned matrix multipiled by a block matrix results in a column partioned matrix -/
+lemma fromColumns_mul_fromBlocks (A₁ : Matrix m m₁ R) (A₂ : Matrix m m₂ R)
+    (B₁₁ : Matrix m₁ n₁ R) (B₁₂ : Matrix m₁ n₂ R) (B₂₁ : Matrix m₂ n₁ R) (B₂₂ : Matrix m₂ n₂ R) :
+    (fromColumns A₁ A₂) ⬝ fromBlocks B₁₁ B₁₂ B₂₁ B₂₂ =
+      fromColumns (A₁ ⬝ B₁₁ + A₂ ⬝ B₂₁) (A₁ ⬝ B₁₂ + A₂ ⬝ B₂₂) := by
+  funext i j
+  rw [fromColumns, fromColumns, fromBlocks]
+  cases j
+  all_goals simp only [of_apply, add_apply, Fintype.sum_sum_type, Sum.elim_inl,
+    Sum.elim_inr, mul_apply]
+
+/-- A block matrix mulitplied by a row partitioned matrix gives a row partitioned matrix -/
+lemma fromBlocks_mul_fromRows (A₁ : Matrix n₁ n R) (A₂ : Matrix n₂ n R)
+    (B₁₁ : Matrix m₁ n₁ R) (B₁₂ : Matrix m₁ n₂ R) (B₂₁ : Matrix m₂ n₁ R) (B₂₂ : Matrix m₂ n₂ R) :
+    fromBlocks B₁₁ B₁₂ B₂₁ B₂₂ ⬝ (fromRows A₁ A₂) =
+      fromRows (B₁₁ ⬝ A₁ + B₁₂ ⬝ A₂) (B₂₁ ⬝ A₁ + B₂₂ ⬝ A₂) := by
+  funext i j
+  rw [fromRows, fromRows, fromBlocks]
+  cases i
+  all_goals simp only [of_apply, add_apply, Fintype.sum_sum_type, Sum.elim_inl, Sum.elim_inr,
+    mul_apply]
+
+end Semiring
+
+section CommRing
+
+variable [CommRing R]
+
+/-- Multiplication of a matrix by its inverse is commutative.
+This is the column and row partitioned matrix form of `Matrix.mul_eq_one_comm`.
+
+The condition `e : n ≃ n₁ ⊕ n₂` states that `fromColumns A₁ A₂` and `fromRows B₁ B₂` are "square".
+-/
+lemma fromColumns_mul_fromRows_eq_one_comm (e : n ≃ n₁ ⊕ n₂)
+    (A₁ : Matrix n n₁ R) (A₂ : Matrix n n₂ R) (B₁ : Matrix n₁ n R) (B₂ : Matrix n₂ n R) :
+    fromColumns A₁ A₂ ⬝ fromRows B₁ B₂ = 1 ↔ fromRows B₁ B₂ ⬝ fromColumns A₁ A₂ = 1 := by
+  constructor
+  intro h
+  rw [← Matrix.submatrix_id_id (fromRows B₁ B₂ ⬝ fromColumns A₁ A₂),
+    ← Matrix.submatrix_mul_equiv (fromRows B₁ B₂) (fromColumns A₁ A₂) _ e.symm _]
+  swap
+  intro h
+  rw [← Matrix.submatrix_id_id (fromColumns A₁ A₂ ⬝ fromRows B₁ B₂),
+    ← Matrix.submatrix_mul_equiv (fromColumns A₁ A₂) (fromRows B₁ B₂) _ e _]
+  all_goals
+  ( rw [mul_eq_one_comm, ← Matrix.submatrix_mul, h, submatrix_one_equiv]
+    exact Function.bijective_id )
```

Comment:
I tried to restructure the proof to emphasise `mul_eq_one_comm` and then rewrite both sides so they'd resemble this. I came up with the following:
```suggestion
  calc fromColumns A₁ A₂ ⬝ fromRows B₁ B₂ = 1
  _ ↔ submatrix (fromColumns A₁ A₂) id e ⬝ submatrix (fromRows B₁ B₂) e id = 1 := by
    simp
  _ ↔ submatrix (fromRows B₁ B₂) e id ⬝ submatrix (fromColumns A₁ A₂) id e = 1 :=
    mul_eq_one_comm
  _ ↔ reindex e.symm e.symm (fromRows B₁ B₂ ⬝ fromColumns A₁ A₂) = reindex e.symm e.symm 1 := by
    simp only [reindex_apply, Equiv.symm_symm, submatrix_one_equiv,
        submatrix_mul (he₂ := Function.bijective_id)]
  _ ↔ fromRows B₁ B₂ ⬝ fromColumns A₁ A₂ = 1 :=
    (reindex _ _).injective.eq_iff
```